### ISSUE TITLE
data race in completer

### ISSUE THREAD

**Issue by slingamn** [ergochat/readline#30]
This race is still present:

https://github.com/chzyer/readline/issues/219

Ctrl-R writes to the search mode status, wrapWriter reads it on asynchronous writes, and these are not necessarily synchronized.

**Comment by slingamn** [comment]
This is partially addressed by #29, but the completer is likely still racey:

https://github.com/ergochat/readline/blob/2513abd5307860ede6de364e5f898496740d9db3/operation.go#L71-L73

**Comment by slingamn** [comment]
1. Start from 717a7d6fd9486e0e2ed9bdb8d495934550f0e499
2. `go build -race example/readline-demo/readline-demo.go`
3. Type `async` to start the async write goroutine
4. Go in and out of tab-complete mode a few times:

```
==================
WARNING: DATA RACE
Write at 0x00c0000cc0f0 by main goroutine:
  github.com/ergochat/readline.(*opCompleter).EnterCompleteMode()
      /data/shivaram/workspace/readline/complete.go:467 +0x4d
  github.com/ergochat/readline.(*opCompleter).OnComplete()
      /data/shivaram/workspace/readline/complete.go:128 +0xa48
  github.com/ergochat/readline.(*operation).readline()
      /data/shivaram/workspace/readline/operation.go:323 +0x5fa
  github.com/ergochat/readline.(*operation).Runes()
      /data/shivaram/workspace/readline/operation.go:430 +0x217
  github.com/ergochat/readline.(*operation).String()
      /data/shivaram/workspace/readline/operation.go:397 +0x47
  github.com/ergochat/readline.(*Instance).ReadLine()
      /data/shivaram/workspace/readline/readline.go:213 +0x27
  github.com/ergochat/readline.(*Instance).Readline()
      /data/shivaram/workspace/readline/readline.go:218 +0x4b6
  main.main()
      /data/shivaram/workspace/readline/example/readline-demo/readline-demo.go:110 +0x4a9

Previous read at 0x00c0000cc0f0 by goroutine 12:
  github.com/ergochat/readline.(*opCompleter).IsInCompleteMode()
      /data/shivaram/workspace/readline/complete.go:137 +0x246
  github.com/ergochat/readline.(*operation).write()
      /data/shivaram/workspace/readline/operation.go:71 +0x205
  github.com/ergochat/readline.(*wrapWriter).Write()
      /data/shivaram/workspace/readline/operation.go:42 +0x84
  github.com/ergochat/readline.(*Instance).Write()
      /data/shivaram/workspace/readline/readline.go:275 +0x8b
  fmt.Fprintf()
      /usr/local/go/src/fmt/print.go:225 +0xb1
  main.asyncSleep()
      /data/shivaram/workspace/readline/example/readline-demo/readline-demo.go:79 +0x1f7
  main.main.func4()
      /data/shivaram/workspace/readline/example/readline-demo/readline-demo.go:177 +0x44

Goroutine 12 (running) created at:
  main.main()
      /data/shivaram/workspace/readline/example/readline-demo/readline-demo.go:177 +0xb31
==================
```

**Comment by slingamn** [comment]
Fixed by #50
